source/common/formatter/substitution_formatter.h

```diff
@@ -253,8 +255,192 @@ template <class FormatterContext> class FormatterBaseImpl : public FormatterBase
+  using Formatters = std::vector<FormatterProviderBasePtr<FormatterContext>>;
+
+  JsonFormatterImplBase(const ProtobufWkt::Struct& struct_format, bool keep_value_type,
+                        bool omit_empty_values, const CommandParsers& commands = {})
+      : omit_empty_values_(omit_empty_values), keep_value_type_(keep_value_type),
+        empty_value_(omit_empty_values_ ? std::string()
+                                        : std::string(DefaultUnspecifiedValueStringView)) {
+    const auto elements = JsonFormatBuilder(keep_value_type).fromStruct(struct_format);
+    for (const auto& element : elements) {
+      if (absl::holds_alternative<JsonFormatBuilder::TmplString>(element)) {
+        elements_.emplace_back(SubstitutionFormatParser::parse<FormatterContext>(
+            absl::get<JsonFormatBuilder::TmplString>(element).value_, commands));
+      } else {
+        elements_.emplace_back(absl::get<JsonFormatBuilder::JsonString>(element).value_);
+      }
+    }
+  }
+
+  std::string formatWithContext(const FormatterContext& context,
+                                const StreamInfo::StreamInfo& info) const {
+    std::string log_line;
+    log_line.reserve(512);
+
+    std::string sanitize_buffer;
+
+    for (const auto& element : elements_) {
+      // 1. Handle the raw string element.
+      if (absl::holds_alternative<std::string>(element)) {
+        log_line.append(absl::get<std::string>(element));
+        continue;
+      }
+      ASSERT(absl::holds_alternative<Formatters>(element));
+
+      const auto& formatters = absl::get<Formatters>(element);
+      ASSERT(!formatters.empty());
+
+      // 2. Handle the formatter element with multiple providers or case
+      //    that value type needs not to be kept.
+      if (formatters.size() > 1 || !keep_value_type_) {
+        // Multiple providers forces string output.
+        log_line.push_back('"');
+        for (const auto& provider : formatters) {
+          const auto bit = provider->formatWithContext(context, info);
+          if (!bit.has_value()) {
+            log_line.append(empty_value_);
+            continue;
+          }
+          log_line.append(Json::sanitize(sanitize_buffer, bit.value()));
+        }
+        log_line.push_back('"');
+
+        continue;
+      }
+
+      // 4. Handle the formatter element with a single provider and value
+      //    type needs to be kept.
+      auto value = formatters[0]->formatValueWithContext(context, info);
+      switch (value.kind_case()) {
+      case ProtobufWkt::Value::KIND_NOT_SET:
+      case ProtobufWkt::Value::kNullValue:
+        log_line.append(JsonFormatBuilder::NullValue);
+        break;
+      case ProtobufWkt::Value::kNumberValue:
+        log_line.append(fmt::to_string(value.number_value()));
+        break;
+      case ProtobufWkt::Value::kStringValue:
+        log_line.push_back('"');
+        log_line.append(Json::sanitize(sanitize_buffer, value.string_value()));
+        log_line.push_back('"');
```

Comment:
use absl::StrCat to append a bunch of fragments to a string.

Reply:
Will it provide better performance when we just need to push a single char to exist string?

Reply:
One call to absl::StrCat would replace these 3 lines and work faster because it would right-size the buffer with a single conditional.

Reply:
let me have a try anyway. :)

Reply:
sorry I meant absl::StrAppend.

Reply:
Yeah, I know, I used that. :)